050-unidirectional-topic-operator.md

```diff
@@ -0,0 +1,801 @@
+>       partitions: 12
+>       replicas: 50
+>     status: 
+>       observedGeneration: 1
+>       topicName: foo
+>       conditions:
+>       - type: Ready
+>         status: True
+>         lastTransitionTime: 20230301T103000Z
+>     ```
+> 
+---
+
+Because topics can be created directly in Kafka we need a mechanism for converting an unmanaged topic to a managed topic, aka "managing" the topic.
+This can easily be done by creating a matching `KafkaTopic`: The operator will attempt the `CreateTopics` request and if it receives a `TOPIC_ALREADY_EXISTS` will proceed with reconciliation as it would for a `KafkaTopic` modification.
+That is, the configuration of the topic in Kafka will be changed to match the `KafkaTopic` (or fail for the same possible reasons and with the same error conditions in the `status` as can happy for any update).
+
+---
+
+#### Example 2: `KafkaTopic` creation when topic exists in Kafka
+
+This is exactly the same as Example 1.
+
+---
+
+### Modification to `KafkaTopics` and timed reconciliation
+
+Via a Kube watch, and also based on a timer, the operator will
+ensure that the state of a managed topic in a Kafka cluster reflects the `spec` of its `KafkaTopic`.
+
+The process will first describe the topic and its configs and then make alterations, if required.
+
+* Modification to topic config will be via `Admin.incrementalAlterConfigs()` using the `SET` operation.
+* Creation of partitions will be supported
+* Deletion of partitions is not supported by Kafka and will result in a suitable `status.condition` in the `KafkaTopic`. 
+    ```yaml
+    status:
+      conditions:
+      - type: Ready
+        status: False
+        reason: NotSupported
+        message: Decrease of spec.partitions is not supported by Kafka
+        lastTransitionTime: 20230301T103000Z
+    ```
+    The user will then need to decide how to proceed, but typically they might just revert the `spec.partitions`.
+* Changes to `spec.replicas` could be supported via CC, but that is not part of _this_ proposal.
+    ```yaml
+    status:
+      conditions:
+      - type: Ready
+        status: False
+        reason: NotSupported
+        message: Changing spec.replicas is not supported by the operator
+        lastTransitionTime: 20230301T103000Z
+    ```
+
+> Note that both these `NotSupported` reasons can arise both from changes being made to the `KafkaTopic`, or changes made directly in Kafka.
+> For example, the UTO cannot tell whether the `spec.partitions` was decreased, or whether it was unchanged and the conflict arises because someone increased the number of partitions directly in Kafka so that it looks like the `KafkaTopic` is requesting a decrease. 
+> In any case, when the user takes action to fix the problem the `KafkaTopic` will (eventually) get reconciled and the `status` will revert to `Ready`.
+
+### Topic deletion
+
+We will use a [Kube finalizer](https://kubernetes.io/blog/2021/05/14/using-finalizers-to-control-deletion/) for deletion.
```

Comment:
The finalizers still make me nervous. Have you considered the possibility to allow users to disable their use and take the risk the topics might not delete in Kafka when the `KafkaTopic` is deleted?

Reply:
I was not really questioning the `.spec.managed` here. I'm wondering if we should have a flag to simply disable the finalizers when the user decides so. I.e. if they would be disabled, they would not be set but the user will accept that there might be situations when deleting the `KafkaTopic` does not trigger the deletion in Kafka.